crates/pallet-executor/src/lib.rs

```diff
@@ -109,10 +117,37 @@ mod pallet {
 
             // TODO: ensure the receipt is ready to be applied
 
-            // TODO: apply the execution receipt.
+            let SignedExecutionReceipt {
+                execution_receipt, ..
+            } = signed_execution_receipt;
 
-            Self::deposit_event(Event::ExecutionReceiptStored {
-                receipt_hash: signed_execution_receipt.hash(),
+            let primary_hash = execution_receipt.primary_hash;
+            let primary_number = execution_receipt.primary_number;
+
+            // Execution receipt starts from the primary block #1.
+            if primary_number > 1u32 {
+                ensure!(
+                    Receipts::<T>::get(primary_number - 1u32).is_some(),
+                    Error::<T>::MissingParentReceipt
+                );
+            }
+
+            // Apply the execution receipt.
+            <Receipts<T>>::insert(primary_number, execution_receipt);
```

Comment:
I think another TODO is needed to remove newer ERs from storage (counter storage map can be useful here to check quickly if there are newer ERs): in case we had fraud proof and got older ER corrected, all newer ERs are implicitly invalid and should be removed.

Reply:
I believe this is part of the TODO revert the execution chain https://github.com/subspace/subspace/pull/427/files#diff-8d1c28c28055fbaa7e46a4dddbe6fa4ddebd330bbdbc82d032d806c089983413R189

Reply:
Location for it is odd, but true, resolving